kubernetes/config/incluster_config.c

```diff
@@ -0,0 +1,165 @@
+#include "incluster_config.h"
+#include <unistd.h>
+#include <sys/stat.h>
+#include <errno.h>
+
+#define SERVICE_HOST_ENV_NAME "KUBERNETES_SERVICE_HOST"
+#define SERVICE_PORT_ENV_NAME "KUBERNETES_SERVICE_PORT"
+#define SERVICE_HTTPS_PREFIX "https://"
+#define SERVICE_TOKEN_FILENAME "/var/run/secrets/kubernetes.io/serviceaccount/token"
+#define SERVICE_CERT_FILENAME "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+#define AUTH_TOKEN_KEY "Authorization"
+#define BEARER_TOKEN_TEMPLATE "Bearer %s"
+#define BEARER_TOKEN_BUFFER_SIZE 1024
+
+static int checkServiceAccountFile(const char *fileName)
+{
+    static char fname[] = "checkServiceAccountFile()";
+
+    if (!fileName) {
+        fprintf(stderr, "%s: The file name must be specified.\n", fname);
+        return -1;
+    }
+
+    struct stat info;
+    if (-1 == stat(fileName, &info)) {
+        switch (errno) {
+        case ENOENT:
+            fprintf(stderr, "%s: The file %s does not exist.[%s]\n", fname, fileName, strerror(errno));
+            return -1;
+        case EACCES:
+            fprintf(stderr, "%s: No permission to read the file %s.[%s]\n", fname, fileName, strerror(errno));
+            return -1;
+        default:
+            fprintf(stderr, "%s: Cannot retrieve the information of file %s.[%s]\n", fname, fileName, strerror(errno));
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+static int setBasePathInCluster(char **pBasePath)
+{
+    static char fname[] = "setBasePathInCluster()";
+
+    const char *service_host_env = getenv(SERVICE_HOST_ENV_NAME);
```

Comment:
We should switch to `secure_getenv` see the recommendations here:

https://www.gnu.org/software/libc/manual/html_node/Environment-Access.html

"General-purpose libraries should always prefer this function over getenv to avoid vulnerabilities if the library is referenced from a SUID/SGID program."

Reply:
It's OK. I have upated it. And I add 
```c
#define _GNU_SOURCE
```
to the source code because it is required by the function ```secure_getenv``` (http://man7.org/linux/man-pages/man3/getenv.3.html)

"Feature Test Macro Requirements for glibc (see feature_test_macros(7)):
    secure_getenv(): _GNU_SOURCE“